File: static/data/mpiheatedplate.py
""" 
This is a program with mpi to solve a weakly scaled problem (i.e. Steady state 
Heat Distribution problem with boundary condition) for square grid (for each 
process) 
"""
import sys
import numpy as np
from mpi4py import MPI
# check number of parameters and read in epsilon

argv = sys.argv
if  len(argv) == 3:
    try:
        epsilon = float( argv[1] )
        filename=argv[2]
        N = 500
    except:
        "Illegal input"
        sys.exit(1)
elif len(argv) == 4 :
    try:
        epsilon = float( argv[1] )
        filename=argv[2]
        N = int( argv[3] )
    except:
        "Illegal input"
        sys.exit(1)
else:
    print'USAGE: <epsilon> <output-file> <Square-Grid-Size(optional)>'
    sys.exit(1)

# Initializing MPI
comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()
nroot = 0
intrvlGlobDiffCheck = 10 # interval for calculating maximum global diff

# Benchmarking
elapsed_time = 0.
comm.Barrier()
elapsed_time = -MPI.Wtime()
localDiff = np.array( [epsilon] )
globalDiff = np.array( [epsilon] )
if rank == nroot :
    print '############ Steady State Heat Distribution Problem #############'
    print 'running until the difference is < ', epsilon
    

#The local array is square
nr = N 
nc = N 

u = np.zeros(  ( nr+2, nc+2 ), dtype=np.double  )
w = np.zeros(  ( nr+2, nc+2 ), dtype=np.double   )

# Set boundary values and compute mean boundary value. 
#This has the ice bath on the top edge, not the bottom edge.
if rank == 0 : 
    u[ 0 , : ] = 0.        # Up
u[ : , 0 ] = 100.      # left
u[ : , nc+1 ] = 100.   # right
if rank == size-1 :
    u[ nr+1 , : ] = 100.   # bottom


#Initialize interior values to the boundary mean.  This value really
#doesn't matter much, it just affects the convergence rate somewhat.
mean = sum( u[:,0] ) + sum( u[:,nr+1] ) + sum( u[0,:] ) + sum( u[nr+1,:] )
mean = mean / ( 4.0 * (N+2) )

u[ 1:nr , 1:nc ] = mean

# setting up the up and down rank for each process
if rank == 0 :
    up = MPI.PROC_NULL
else :
    up = rank - 1
    
if rank == size - 1 :
    down = MPI.PROC_NULL
else :
    down = rank + 1

# Compute steady-state solution
iterations = 0
tag=0

while ( globalDiff[0] >= epsilon ) :
    # sending up and receiving down
    comm.Sendrecv(  [  u[1, 1:nc+1], MPI.DOUBLE  ], up, tag, [  u[nr+1, 1:nc+1], MPI.DOUBLE  ], down , tag )
    # sending down and receiving up
    comm.Sendrecv(  [  u[ nr , 1:nc+1 ] , MPI.DOUBLE ] , down, tag, [  u[ 0, 1:nc+1 ], MPI.DOUBLE  ], up , tag )
    # calculating average
    w[ 1:-1, 1:-1 ] = 0.25 * ( u[:-2,1:-1] +u[2:,1:-1] + u[1:-1,:-2] + u[1:-1,2:] )
    
    localDiff[0] = np.max(  np.abs( w[1:-1,1:-1] - u[1:-1,1:-1] )  )
    
    if iterations % intrvlGlobDiffCheck == 0 :
        comm.Allreduce( localDiff, globalDiff, op = MPI.MAX  )
        #if rank == nroot :
        #    print "Running with iterations :", iterations, 'and global diff : ', globalDiff[0]
    
    if globalDiff[0] >= epsilon :
        u[ 1:N+1 , 1:N+1 ]  =  w[ 1:N+1 , 1:N+1 ].copy() 
    iterations += 1
if rank == nroot :        
    print 'completed; iterations = ', iterations

# Write solution to output file
filename = filename + str( rank ) + ".txt"
fout = open (filename,'w')
if rank == nroot :
    print  >> fout, N, N, size
for i in range( 1 , N+1 ) :
    line = " ".join(   map(  str , list( u[i, 1:N+1] )  )   )
    row = line + "\n"
    fout.write( row )

# All done!
print  "Process ", rank, " : wrote output file ", filename
# Benchmarking
elapsed_time += MPI.Wtime()

if rank == nroot :
    print 'Elapsed Time : ', elapsed_time, ' sec'

File: static/data/mpiheatedplate.cxx
/* Parallel Solution to Steady-State Heat Problem */

#include <iostream>
#include <cmath>
#include <ctime>
#include <string>
#include <sstream>
#include <fstream>
#include <iomanip>
#include <string>
#include <mpi.h>

//If your N is too large, it will have problems allocating the necessary memory
#define N 500

#define MAX_ITER 100000000

double topBC=100;
double bottomBC=0.;
double edgeBC=100.;

void set_bcs(double **u, int nr, int nc, int rank, int npes);

using namespace std; 

int main (int argc, char *argv[]) {
  int i, j;
  int iterations = 0;
  double epsilon;
  double diff;            // Change in value 
  float time, etime;

  // Added for MPI
  int nr, nrl, nc;
  int rank, npes;
  MPI_Status status;
  int root=0, tag=0;
  int up, down;
  double gdiff;
  string fname;

  //Initialize MPI
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD,&npes);
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);

  // check number of parameters and read in epsilon 
  if (argc != 3) {
    if (rank==root) {
       printf ("USAGE:  %s <epsilon> <output-file>\n", argv[0]);
    }
    MPI_Finalize();
    return 1;
  }
  stringstream ssepsilon;
  ssepsilon<<argv[1];
  if (ssepsilon.fail()) {
    printf ("Error reading in epsilon value\n");
    return 2;
  }

  ssepsilon>>epsilon;
  if (rank==root) {
     cout<<"running until the difference is <="<<epsilon<<"\n";
  }

  diff = epsilon;

  //Usually we would read in nr and nc; they represent the global grid RxC;
  //For this case they are both N.
  nr=N;
  nc=N;

  //Weak scaling
  //nrl=nr;

  //Strong scaling
  if (nr%npes!=0) {
     MPI_Finalize();
     cout<<"Not an even division of the arrays.\n";
     return 3;
  }
  else {
     nrl=nr/npes;
  }

  //Find my neighbors
  if (rank==0) {
      up=MPI_PROC_NULL;
  }
  else {
      up=rank-1;
  }

  if (rank==npes-1) {
      down=MPI_PROC_NULL;
  }
  else {
      down=rank+1;
  }

  int nrows=nrl+2;
  int ncols=nc+2;
  double **u=new double*[nrows];  // Old values
  double **w=new double*[nrows];  // New values

  double *uptr=new double[(nrows)*(ncols)];
  double *wptr=new double[(nrows)*(ncols)];

  for (i=0;i<(nrl+2);++i,uptr+=ncols) 
     u[i] = uptr;
  for (i=0;i<(nrl+2);++i,wptr+=ncols) 
     w[i] = wptr;

  //Compute a mean of the boundary values as a first guess
  double mean=(2.*edgeBC+topBC+bottomBC)/4.0;

  // Initialize all values to mean
  for ( i = 0; i <= nrl+1; i++ )
    for (j = 0; j <= nc+1; j++ )
       u[i][j] = mean;
      

  //Set physical boundary conditions
  set_bcs(u,nrl,nc,rank,npes);

  //Walltime from process 0 is good enough for me 
  if (rank==root) { 
     time=MPI_Wtime();
  }

  // Compute steady-state solution 
  // Have to remove the condition on diff to keep processes in sync

  while ( iterations<MAX_ITER ) {
    diff = 0.0;  //reset to get maximum
    for (i=1; i<=nrl;i++)
      for (j=1;j<=nc;j++) {
          w[i][j] = (u[i-1][j] + u[i+1][j] + u[i][j-1] + u[i][j+1])/4.0;
    	  if ( abs(w[i][j] - u[i][j]) >= diff )
	     diff = abs(w[i][j] - u[i][j]);
      }

      //Exchange halo values (one ghost row each side)
      MPI_Sendrecv(&w[1][1],nc, MPI_DOUBLE,up,tag,&w[nrl+1][1],
                            nc, MPI_DOUBLE,down,tag,MPI_COMM_WORLD,&status);

      MPI_Sendrecv(&w[nrl][1],nc,MPI_DOUBLE,down,tag,&w[0][1],
                              nc,MPI_DOUBLE,up,tag,MPI_COMM_WORLD,&status);

      //Find max of dff in all the processors.
      MPI_Allreduce(&diff,&gdiff,1,MPI_DOUBLE,MPI_MAX,MPI_COMM_WORLD);

      if (gdiff <= epsilon) {
          break;
      }
      for (i=0; i<=nrl+1;i++) 
         for (j=0;j<=nc+1;j++)
            u[i][j] = w[i][j];

      set_bcs(u,nrl,nc,rank,npes);

      iterations++;
  } 

  if (rank==root) {
      cout<<"completed; iterations = "<<iterations<<" in "<<MPI_Wtime()-time<<"\n";
  }

  // Write solution to output file
  ofstream fp;
  fname+=argv[2]+to_string(rank);
  fp.open(fname,ios::out);
  for (i = 1; i <= nrl; i++) {
    for (j = 1; j <= nc; j++)
      fp<<u[i][j]<<" ";
    fp<<"\n";
  }
  fp.close();

  // All done!
  cout<<"wrote output file "<<fname<<"\n";

  MPI_Finalize();
}

void set_bcs(double **u, int nr, int nc, int rank, int npes) {
  /* Set boundary values.
   * This has an ice bath on the top edge.
   * Note: when plotting, 0,0 is the bottom.
   */

  if (rank==0) {
     for (int i=0;i<=nc+1;++i){
        u[0][i]=bottomBC;
     }
  }
  if (rank==npes-1) {
      for (int i=0;i<nc+1;++i){
        u[nr+1][i]=topBC;
      }
  }

  for (int i=0;i<=nr+1;++i){
      u[i][0]=edgeBC;
      u[i][nc+1]=edgeBC;
  }
}
